### PR TITLE
[spec] add partial specification for index_of function for big_vector

### PR DESCRIPTION
**Pull request by aalok-t**
### Description

Add functional specification for the big_vector module. If `index_of(v, val)` returns `(true, i)`, then `val` must occur in `v` at index `i`.

### CODE REVIEW

File: aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.spec.move [hunk 1]
```diff
@@ -100,20 +117,14 @@ spec aptos_std::big_vector {
...
-
     spec reverse<T>(v: &mut BigVector<T>) {
         pragma verify=false;
+        // ensures forall i in 0..v.end_index: spec_at(v, i) == spec_at(old(v), v.end_index - (i + 1));
     }
 
     spec index_of<T>(v: &BigVector<T>, val: &T): (bool, u64) {
-        pragma verify=false;
+        ensures (result_1 == true) ==> (spec_at(v, result_2) == val);
+        // ensures (result_1 == false) ==> (forall i in 0..v.end_index: spec_at(v, i) != val);
```
Comment (aalok-t) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.spec.move:127]: Proving this assertion will complete the specification for the index_of function. Writing the loop invariant for this specification is challenging.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.spec.move [hunk 2]
```diff
@@ -65,6 +69,12 @@ spec aptos_std::big_vector {
         ensures result == spec_at(v, i);
     }
 
+    spec append<T: store>(lhs: &mut BigVector<T>, other: BigVector<T>) {
+        pragma verify=false;
+        // ensures forall i in 0..old(lhs).end_index: spec_at(old(lhs), i) == spec_at(lhs, i);
```
Comment (aalok-t) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.spec.move:74]: Comments are left throughout the file in order to note the formulated specifications. Proving each of these specifications individually would be the next step.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.move [hunk 3]
```diff
@@ -81,7 +81,10 @@ module aptos_std::big_vector {
         let other_len = length(&other);
         let half_other_len = other_len / 2;
         let i = 0;
-        while (i < half_other_len) {
+        while ({spec {
+            invariant true;
```
Comment (rahxephon89) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.move:85]: Please remove this

**Review by rahxephon89 — CHANGES_REQUESTED** [review]

File: aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.spec.move (hunk 2 above)
Comment (rahxephon89) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.spec.move:74]: Maybe you could add another comment explaining that it is not proved but a work to be done next?

File: aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.move [hunk 4]
```diff
@@ -138,6 +138,8 @@ module aptos_std::big_vector {
             spec {
                 invariant cur_bucket_index <= num_buckets;
                 invariant table_with_length::spec_len(v.buckets) == num_buckets;
+                invariant cur_bucket_index > 0;
+                // assert i != 0 ==> (spec_at(v, 0) == spec_at(old(v), 0));
```
Comment (rahxephon89) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.move:142]: Maybe remove this commented spec for now in this PR

**Pull request closed without merging**